### PR TITLE
VLM performance metrics.

### PR DESCRIPTION
**Pull request by popovaan**
VLM performance metrics.
Ticket: CVS-156661

### CODE REVIEW

File: src/cpp/src/visual_language/pipeline.cpp [hunk 1]
```diff
@@ -149,10 +158,27 @@ class ov::genai::VLMPipeline::VLMPipelineImpl {
             m_language.reset_state();
             m_language.get_tensor("attention_mask").set_shape({1, 0});
         }
+        auto generate_end_time = std::chrono::steady_clock::now();
+        decoded.perf_metrics = encoded_result.perf_metrics;
+
+        // Common perf metrics
+        auto& res_raw_counters = decoded.perf_metrics.raw_metrics;
+        decoded.perf_metrics.load_time = m_load_time_ms;
+        res_raw_counters.generate_durations = std::vector<MicroSeconds>();
```
Comment (Wovchena) [inline src/cpp/src/visual_language/pipeline.cpp:167]:
```suggestion
```
It should already be default constructed

**Review by Wovchena — APPROVED** [review]

**Comment by ilya-lavrenov** [comment]
@Wovchena @popovaan do we need similar sample https://github.com/openvinotoolkit/openvino.genai/blob/master/samples/cpp/benchmark_genai/benchmark_genai.cpp ? to show VLM perf metrics breakdown

File: src/cpp/src/visual_language/inputs_embedder.cpp [hunk 2]
```diff
@@ -291,6 +294,10 @@ class InputsEmbedderMiniCPM : public InputsEmbedder::IInputsEmbedder {
             m_image_id = 0;
         }
```
Comment (ilya-lavrenov) [inline src/cpp/src/visual_language/inputs_embedder.cpp:295]: I see `resample(...)` in the loop, which is embedding part IMO 

@Wovchena how do you think? I expect that tokenization should imply merging final embeddings only

**Comment by ilya-lavrenov** [comment]
Please, rebase to latest master to resolve issue with Whisper https://github.com/openvinotoolkit/openvino.genai/actions/runs/12054144518/job/33690367538?pr=1263

**Comment by Wovchena** [comment]
> @Wovchena @popovaan do we need similar sample https://github.com/openvinotoolkit/openvino.genai/blob/master/samples/cpp/benchmark_genai/benchmark_genai.cpp ? to show VLM perf metrics breakdown

No need. The difference with the base sample is in one field. Another question is if we need to extend the current sample. I belie no, because there's a dedicated sample. We can update every sample to present perf metrics and that would eliminate benhcmark sample.

File: src/cpp/src/visual_language/inputs_embedder.cpp (hunk 2 above)
Comment (Wovchena) [reply src/cpp/src/visual_language/inputs_embedder.cpp:295]: Yeah, but why is merging should be part of tokenization? I think the measurement should happen in `get_encoded_input_ids()`

**Comment by ilya-lavrenov** [comment]
> > @Wovchena @popovaan do we need similar sample https://github.com/openvinotoolkit/openvino.genai/blob/master/samples/cpp/benchmark_genai/benchmark_genai.cpp ? to show VLM perf metrics breakdown
> 
> No need. The difference with the base sample is in one field. Another question is if we need to extend the current sample. I belie no, because there's a dedicated sample. We can update every sample to present perf metrics and that would eliminate benhcmark sample.

IMO, approaches should be similar for VLM and LLM.
We either need to drop current perf sample for LLM or add new one for VLM.

Personally, I don't mind to have as many samples as possible to avoid having one sample/tool which can demonstrate everything.

File: src/cpp/src/visual_language/inputs_embedder.cpp (hunk 2 above)
Comment (Wovchena) [reply src/cpp/src/visual_language/inputs_embedder.cpp:295]: Not really. I was thinking about ignoring that stage :)

**Comment by ilya-lavrenov** [comment]
@popovaan could you please provide example output in a format similar to LLM perf sample?

**Comment by Wovchena** [comment]
VLM and Whisper are more alike than VLM and LLM in context of benchmarking because Whisper has extra metrics just like VLM. And Whisper doesn't have such sample

File: src/cpp/src/visual_language/inputs_embedder.cpp (hunk 2 above)
Comment (ilya-lavrenov) [reply src/cpp/src/visual_language/inputs_embedder.cpp:295]: what is the goal to break down 1st token at all and expose it partially (e.g. only image embedding), but not the whole breakdown?

As a user, I would expect either:
- sum up all stages in 1st token and it will result in 1 token latency
- hide all details about 1st token at all, as they are useless (?)

Comment (Wovchena) [reply src/cpp/src/visual_language/inputs_embedder.cpp:295]: the whole breakdown was considered as too detailed not needed report. But I don't know why we need VLM unique stage as a separate metric

Comment (ilya-lavrenov) [reply src/cpp/src/visual_language/inputs_embedder.cpp:295]: agreed on meeting, that we need to measure whole prompt embedding preparation stage which includes text tokenization + embedding, image embeddings and final merging.

**Comment by ilya-lavrenov** [comment]
> @popovaan could you please provide example output in a format similar to LLM perf sample?

this can be added via sample. I suppose it's easy to copy existing LLM sample and adjust a few lines

**Comment by popovaan** [comment]
> @popovaan could you please provide example output in a format similar to LLM perf sample?

Load time: 1982.00 ms
Generate time: 13820.99 ± 64.62 ms
Tokenization time: 1.26 ± 0.09 ms
Detokenization time: 0.33 ± 0.05 ms
Embeddings preparation time: 5733.85 ± 26.34 ms
TTFT: 11246.98 ± 80.55 ms
TPOT: 135.45 ± 4.73 ms/token 
Throughput: 7.38 ± 0.26 tokens/s

File: samples/python/benchmark_vlm/benchmark_vlm.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    parser.add_argument("-p", "--prompt", type=str, default="The Sky is blue because", help="Prompt")
+    parser.add_argument("-i", "--image", type=str, default="image.jpg", help="Image")
+    parser.add_argument("-nw", "--num_warmup", type=int, default=1, help="Number of warmup iterations")
+    parser.add_argument("-n", "--num_iter", type=int, default=2, help="Number of iterations")
+    parser.add_argument("-mt", "--max_new_tokens", type=int, default=20, help="Maximal number of new tokens")
+    parser.add_argument("-d", "--device", type=str, default="CPU", help="Device")
+
+    args = parser.parse_args()
+
+    # Perf metrics is stored in DecodedResults.
```
Comment (ilya-lavrenov) [inline samples/python/benchmark_vlm/benchmark_vlm.py:37]: VLMDecodedResults  ?

**Review by ilya-lavrenov — APPROVED** [review]

File: samples/cpp/benchmark_vlm/README.md
Comment (ilya-lavrenov) [inline samples/cpp/benchmark_vlm/README.md:1]: it would be great if you could merge all VLM samples to a single directory with common README (generic info and per-sample sections similar to text2image)

File: samples/python/benchmark_vlm/benchmark_vlm.py (hunk 3 above)
Comment (popovaan) [reply samples/python/benchmark_vlm/benchmark_vlm.py]: Yes, fixed.

File: samples/cpp/benchmark_vlm/README.md
Comment (popovaan) [reply samples/cpp/benchmark_vlm/README.md:1]: Done.
